MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs

```diff
@@ -1235,32 +1235,53 @@ private void OnSetAnimationState(SetAnimationState payload, Action onCompleteCal
             onCompleteCallback?.Invoke();
         }
 
-        [CommandHandler(typeof(SetSoundState))]
-        private void OnSetSoundState(SetSoundState payload, Action onCompleteCallback)
+        [CommandHandler(typeof(SetMediaState))]
+        private void OnSetMediaState(SetMediaState payload, Action onCompleteCallback)
         {
-            if (payload.SoundCommand == SoundCommand.Start)
+            if (payload.MediaCommand == MediaCommand.Start)
             {
+                if (_mediaInstances == null)
+                {
+                    _mediaInstances = new Dictionary<Guid, System.Object>();
+                }
                 AudioSource soundInstance = App.SoundManager.TryAddSoundInstance(this, payload.Id, payload.SoundAssetId, payload.Options, payload.StartTimeOffset);
                 if (soundInstance)
                 {
-                    if (_soundInstances == null)
+                    _mediaInstances.Add(payload.Id, soundInstance);
+                }
+                else
+                {
+                    var factory = MREAPI.AppsAPI.VideoPlayerFactory
+                        ?? throw new ArgumentException("Cannot movie player not implemented library.");
+                    IVideoPlayer videoPlayer = factory.CreateVideoPlayer(this);
+
+                    var videoStreamDescription = MREAPI.AppsAPI.AssetCache.GetAsset(payload.SoundAssetId) as VideoStreamDescription;
+                    if (videoStreamDescription != null)
                     {
-                        _soundInstances = new Dictionary<Guid, AudioSource>();
+                        videoPlayer.Play(videoStreamDescription, payload.Options, payload.StartTimeOffset);
                     }
-                    _soundInstances.Add(payload.Id, soundInstance);
+                    _mediaInstances.Add(payload.Id, videoPlayer);
                 }
             }
             else
             {
-                if (_soundInstances != null && _soundInstances.TryGetValue(payload.Id, out AudioSource soundInstance))
+                if (_mediaInstances != null && _mediaInstances.TryGetValue(payload.Id, out System.Object mediaInstance))
```

Comment:
cslint: Combine into `else if` to reduce indentation.

Reply:
I refactored this into a cleaner 3-level switch statement, so fixed.

---
In reply to: [302635996](https://github.com/Microsoft/mixed-reality-extension-unity/pull/86#discussion_r302635996) [](ancestors = 302635996)